### ISSUE TITLE
[Python] [sklearn] Improvements to Sklearn Python interface

### ISSUE THREAD

**Issue by mlisovyi** [VowpalWabbit/vowpal_wabbit#1469]
Hi,

I'm just starting to use VW in python. I'm very happy that there is a python binding and one dores not need to go around to dump `.vw` files with custom functions, running from a command line and then reading in the predictions in python again. And in particular, the sklearn wrapper is very useful, as it allows to use multiple handy functionalities from that package and makes usage universal across many algorithms. So thanks a lot!

I have a few questions to this interface:
1. Is there a documentation of the API (pyvw in general, and sklearn_vw in particular)? Would it be possible to add auto-generated docs from the class and function description? For example, one can use [Sphinx](http://www.sphinx-doc.org/en/master/) for that or any other tool that you would fancy. This will help users to find information faster.
2. I wanted to understand how exactly training is handled by the sklearn interface. It seems that there are 2 regimes: 
2.1 VW native approach based on a `.vw` file. This seems to be enabled by the `data` arg in the constructor and learning done in the `pyvw.vw()` instantiation in `get_vw()`. Is that understanding correct? If so, this is very nice as it allows online training with a restricted memory footprint, as advertised by the tool. However, from [here](https://github.com/JohnLangford/vowpal_wabbit/blob/master/python/vowpalwabbit/sklearn_vw.py#L251), it seems that even in this case one runs one (and always only one) _additional_ pass over the data provided to the `.fit()` function. Would it make sense to set the number of passes to 0 in this case?
2.2 The sklearn-like interface based on in-memory X and y arrays. In this case one iterates over `pyvw.vw.learn()` calls. That's fine for small datasets, I imagine. It also makes start of VW usage for people with sklearn experience much faster and smoother. The only thing here, would it be possible to add an additional parameter to the constructor that would be specific to the sklearn API and would allow one to convert `[0,1]` labels in the classification into `[-1,1]`. This simplifies life a lot, since in most cases data have `0` label instead of `-1`. One can maybe even reuse the existing `convert_to_vw_` attribute to trigger this conversion, but this would not be backward-compatible (I'm not sure how many people use the sklearn interface of VW by now).

Related to the approach discussed in 2.1, why is `fit_` attritube set to `True` already in the constructor in this case? It seems that the actual fit is done still inside `get_vw()` call from `.fit()`. I might have got it wrong, but if I'm right, than this defies the definition of the `fit_` attribute, as it leads to actual model fit inside `.predict()` call, which would make this call slow and confuse a user, is `.fit()` was never called.

I'm happy to help with implementation of some of the issues discussed above, if you find them useful.

Cheers,
Misha

**Comment by JohnLangford** [comment]
Maybe @Scott-Graham-Bose can respond?  Certainly, your help in better documentation is welcome.

**Comment by mlisovyi** [comment]
I'm happy to help with the docs, but I would first want to hear from the developers to understand the intention and philosophy behind the current version

**Comment by Scott-Graham-Bose** [comment]
Thanks for the detailed questions Misha.

1. There are not currently any docs for either pyvw.py or sklearn_vw.py, and I agree the lowest hanging fruit would be to use sphinx api-doc to autogenerate some docs as most of the functions have at least some doc strings. We can work on adding add'l docs as needed.

2.1 Looking more closely at this (some of it has evolved since I've touched it) I'm not really sure why there's a separate get_vw() function constructing the vw_ object? But, you're right that training currently only happens when you call fit (or predict). I think it would be more intuitive to have the model train immediately if you provide it a data file in the constructor and use the number of passes provided. The assumption is if the user is providing external data they are not also going to call fit. We should add some documentation to let users know if they want to load both an external file and call fit the passes parameter may need to be adjusted manually. This would require a small change to the constructor in the section you referenced to remove the passes override and actually instantiate the self.vw_ object (and avoid setting it to None afterwards of course).

2.2 This seems like a useful feature, I agree triggering this by convert_to_vw makes sense. I would suggest an additional parameter 'convert_labels' which would be passed to the tovw() function which could change the labels to [-1, 1] automatically. This could default to false so it would retain backwards compatibility.

Happy to provide review for any pull requests on these items.

scott

**Comment by mlisovyi** [comment]
@Scott-Graham-Bose thanks a lot for the feedback. I think, i can implement 2.1 and 2.2 along the lines of your proposal, but the docs auto-generation is something where i do not have required expertise,  unfortunately

**Comment by jackgerrits** [comment]
Hi @mlisovyi,

Please see here for hosted generated Python docs: https://vowpalwabbit.github.io/docs/vowpal_wabbit/python/build/html/

Did you decide to have a crack at 2.1 or 2.2? Please feel free to open a PR for them!

@Scott-Graham-Bose are points 2.1 and 2.2 still relevant? I want to make sure since this issue has been around a while.

**Comment by jackgerrits** [comment]
To summarize there are a few items left outstanding here. Anyone should feel free to submit a PR for these:
1. Move vw_ object creation to constructor of sklearn_vw.VW object and remove lazy creation in `fit`, etc calls
2. Allow for passing an object file and number of passes and train immediately on creation of the object
3. Fix documentation to explain if they want to load both an external file and call fit the passes parameter may need to be adjusted manually
4. Add additional parameter (`convert_labels`) that is passed to `tovw()` when `convert_to_vw` is used which would convert  'convert_labels'
    - This would convert labels of the form `[0,1]` to `[-1,1]`
    - This could default to false so it would retain backwards compatibilit

**Comment by mlisovyi** [comment]
Sorry to the delay. I will have a look over the weekend

**Comment by jackgerrits** [comment]
Not a problem at all, thanks! :)

**Comment by gramhagen** [comment]
@jackgerrits I dug in a bit more on item 1. I believe lazy initialization is the right construct for this estimator, and it was added to support sklearn [pipelines](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html#sklearn.pipeline.Pipeline). We should keep the vw model construction out of __init__ to maintain support for this. Handling the case where a data file is provided makes this a little funky, but I think it's better to stick to a single conceptual model, where training only happens when fit is called (and the vw model only needs to be instantiated at this point).

**Comment by Sharad24** [comment]
@jackgerrits Is this safe to close?

**Comment by jackgerrits** [comment]
Closing as it seems like the asks where implemented.
